api/types/templates.go

```diff
@@ -42,3 +42,30 @@ type TemplateServer struct {
 
 // TemplateScriptCredentials stores credentials to servers
 type TemplateScriptCredentials interface{}
+
+// cookbookVersionsMap stores template cookbook versions map
+type cookbookVersionsMap map[string]cookbookVersionsFields
+
+// cookbookVersionsFields stores cookbook versions fields: version/version_id
+type cookbookVersionsFields struct {
+	Version          string `json:"version,omitempty" header:"VERSION" show:"nolist,noshow"`
+	VersionId        string `json:"version_id,omitempty" header:"VERSION_ID" show:"nolist,noshow"`
+	VersionComposite string `json:"omitempty" header:"VERSION_COMPOSITE" show:"nolist,noheader"`
+}
+
+// FillInCookbookVersion resolves adequate cookbook version from version_id
+func (cbf *Template) FillInCookbookVersion(VersionByVersionID map[string]string) {
+	for name, cb := range cbf.CookbookVersions {
+		t := new(cookbookVersionsFields)
+		if _, found := VersionByVersionID[cb.VersionId]; found {
+			cb.Version = VersionByVersionID[cb.VersionId]
+			t.Version = cb.Version
+			t.VersionId = cb.VersionId
+			t.VersionComposite = strings.Join([]string{name, cb.Version}, ":")
+		} else {
+			t.Version = cb.Version
+			t.VersionComposite = strings.Join([]string{name, strings.Replace(cb.Version, " ", "", -1)}, "")
+		}
+		cbf.CookbookVersions[name] = *t
+	}
```

Comment:
Why make n allocations of cookbookVersionsField on the heap when you can use the already allocated value?
```Golang
func (t *Template) FillInCookbookVersionComposite(customCookbookVersionsByVersionID map[string]string) {
	for name, cv := range t.CookbookVersions {
		if v, found := VersionByVersionID[cv.VersionId]; found {
			cv.Version = v
			cv.VersionComposite = strings.Join([]string{name, v}, ":")
		} else {
			cv.VersionComposite = strings.Join([]string{name, strings.Replace(cv.Version, " ", "", -1)}, "")
		}
	}
```
If this does not work, you may need to make cookbookVersionsMap into `map[string]*cookbookVersionsFields`

Reply:
Done!